arch/Alpha/AlphaInstPrinter.c

```diff
@@ -0,0 +1,102 @@
+#ifdef CAPSTONE_HAS_ALPHA
+
+#include <platform.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "../../utils.h"
+#include "AlphaInstPrinter.h"
+#include "AlphaMapping.h"
+#include "../../MCInstPrinter.h"
+
+static const char *getRegisterName(unsigned RegNo);
+
+static void printInstruction(MCInst *, uint64_t, SStream *);
+
+static void printOperand(MCInst *MI, int OpNum, SStream *O);
+
+void Alpha_post_printer(csh ud, cs_insn *insn, char *insn_asm, MCInst *mci)
+{
+	/*
+	   if (((cs_struct *)ud)->detail != CS_OPT_ON)
+	   return;
+	 */
+}
+
+#define GET_INSTRINFO_ENUM
+
+#include "AlphaGenInstrInfo.inc"
+
+#define GET_REGINFO_ENUM
+
+#include "AlphaGenRegisterInfo.inc"
+
+static inline void fill_alpha_register(MCInst *MI, uint32_t reg)
+{
+	if (!(MI->csh->detail == CS_OPT_ON && MI->flat_insn->detail))
+		return;
+	cs_alpha *alpha = &MI->flat_insn->detail->alpha;
+	alpha->operands[alpha->op_count].type = ALPHA_OP_REG;
+	alpha->operands[alpha->op_count].reg = reg;
+	alpha->op_count++;
+}
+
+static inline void fill_alpha_imm(MCInst *MI, int32_t imm)
+{
+	if (!(MI->csh->detail == CS_OPT_ON && MI->flat_insn->detail))
+		return;
+	cs_alpha *alpha = &MI->flat_insn->detail->alpha;
+	alpha->operands[alpha->op_count].type = ALPHA_OP_IMM;
+	alpha->operands[alpha->op_count].imm = imm;
+	alpha->op_count++;
+}
+
+static void printOperand(MCInst *MI, int OpNum, SStream *O)
+{
+	MCOperand *Op;
+	if (OpNum >= MI->size)
+		return;
+
+	Op = MCInst_getOperand(MI, OpNum);
+	if (MCOperand_isReg(Op)) {
+		unsigned reg = MCOperand_getReg(Op);
+		SStream_concat(O, "%s", getRegisterName(reg));
+		fill_alpha_register(MI, reg);
+	} else if (MCOperand_isImm(Op)) {
+		int64_t Imm = MCOperand_getImm(Op);
+		if (Imm >= 0) {
+			if (Imm > HEX_THRESHOLD)
+				SStream_concat(O, "0x%" PRIx64, Imm);
+			else
+				SStream_concat(O, "%" PRIu64, Imm);
+		} else {
+			if (Imm < -HEX_THRESHOLD)
+				SStream_concat(O, "-0x%" PRIx64, -Imm);
+			else
+				SStream_concat(O, "-%" PRIu64, -Imm);
+		}
+
+		fill_alpha_imm(MI, (int32_t)Imm);
+	}
+}
+
+#define PRINT_ALIAS_INSTR
+
+#include "AlphaGenAsmWriter.inc"
+
+const char *Alpha_getRegisterName(csh handle, unsigned int id)
+{
+#ifndef CAPSTONE_DIET
+	return getRegisterName(id);
+#else
+	return NULL;
+#endif
+}
+
+void Alpha_printInst(MCInst *MI, SStream *O, void *Info)
```

Comment:
Please call this function `Alpha_LLVM_printInst(...)` and define it in a separated header `AlphaLinkage.h`. And do the same for the other static LLVM functions which are called from the `AlphaMapper` component (`getInstruction`, `getRegisterName`).

`Alpha_LLVM_printInst` should be called by `AlphaMapping.c::Alpha_printer()`. `Alpha_printer()` should be set as printer in `AlphaModule`.

So the call path is:
`CS` -> `csh->printer()` = `AlphaMapping.c::Alpha_printer()` -> `AlphaInstPrinter.c::Alpha_LLVM_printInst()` -> `AlphaInstPrinter.c::printInst()`

The point of this is that we have only a single function (`Alpha_LLVM_printInst()`) which does nothing else but link the Capstone to the LLVM code. And if you have to manipulate anything on the LLVM data, you can do it in `AlphaMapping.c::Alpha_printer()`. This keeps the `AlphaInstPrinter.c` file relatively close to its C++ equivalent.
And we still have a place to add our code (in `AlphaMapping.c::Alpha_printer()`).

You can check out how it is done in `ARM` and the other auto-sync archs.

Reply:
Done for `Alpha_getRegisterName` and `Alpha_printInst`